propolis/src/inventory.rs

```diff
@@ -1,63 +1,714 @@
+    }
+
+    /// Registers a new entity with the inventory.
+    ///
+    /// Returns an error if the object has already been registered.
+    /// Returns an error if the parent is not registered.
+    pub fn register<T: Entity>(
+        &mut self,
+        parent_id: EntityID,
+        ent: Arc<T>,
+        name: String,
+    ) -> Result<EntityID, RegistrationError> {
+        let rec = Record::new(Some(parent_id), ent, name);
+
+        let id = self.next_id();
+
+        self.insert(rec, id)?;
+        self.insert_into_parent(parent_id, id)?;
+
+        Ok(id)
+    }
+
+    /// Access the concrete type of an entity by ID.
+    ///
+    /// Returns the entity if it exists and has the requested concrete type.
+    pub fn get_concrete<T: Entity>(&self, id: EntityID) -> Option<Arc<T>> {
+        self.entities.get(&id)?.concrete()
+    }
+
+    /// Access an entity's record by ID.
+    pub fn get(&self, id: EntityID) -> Option<&Record> {
+        let record = self.entities.get(&id)?;
+        Some(record)
+    }
+
+    /// Removes an entity from the inventory.
+    ///
+    /// No-op if the entity does not exist.
+    pub fn deregister(
+        &mut self,
+        id: EntityID,
+    ) -> Result<(), RegistrationError> {
+        // Remove all children recursively.
+        let children = {
+            let record = self
+                .entities
+                .get_mut(&id)
+                .ok_or_else(|| RegistrationError::MissingEntity(id))?;
+            std::mem::take(&mut record.children)
+        };
+        for child in &children {
+            self.deregister(*child)?;
+        }
+
+        // Remove the current entity.
+        let record = self.entities.remove(&id).unwrap();
+        let entptr = record.ent.object_id();
+        self.reverse.remove(&entptr);
+
+        // If this entity exists in the parent, remove it.
+        //
+        // Note that the parent may not exist, or it may have already been
+        // removed as a part of the recursive walk.
+        if let Some(parent) = record.parent {
+            let parent = self.entities.get_mut(&parent).unwrap();
+            parent.children.remove(&id);
+        }
+
+        // If this is the root, deregister it too.
+        if let Some(root_id) = self.root.as_ref() {
+            if *root_id == id {
+                self.root = None;
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Returns true if the inventory is empty.
+    pub fn is_empty(&self) -> bool {
+        self.entities.is_empty()
+    }
+
     fn next_id(&mut self) -> EntityID {
         self.next_id += 1;
         EntityID { num: self.next_id }
     }
+
+    fn insert_into_parent(
+        &mut self,
+        parent_id: EntityID,
+        child_id: EntityID,
+    ) -> Result<(), RegistrationError> {
+        let parent = self
+            .entities
+            .get_mut(&parent_id)
+            .ok_or_else(|| RegistrationError::MissingParent(parent_id))?;
+        parent.children.insert(child_id);
```

Comment:
Does this need to return an `Err` if the child is already registered? The behavior on the set won't matter, and this function is internal, but it might help uphold the invariants of the public API.

Reply:
The ordering (in the public api) guarantees that any conflict (e.g., child already registered) will return an error before getting here - I don't really want to return an error, because that error handling code would be impossible to get to.

I could change this to an assertion? `assert!(parent.children.insert(child_id))` would demand that invariant, but IMO also make it more clear that it should never be reachable from the public API

Reply:
Assert sounds good to me -- I agree handling an error gracefully here is both difficult and not really correct. But having the assert would be nice if we change the API or make some other mistake.